### DISCUSSION TITLE
How can I enable adaptive mesh refinement with pyAMReX?

### DISCUSSION THREAD

**Discussion by adamdarx** [AMReX-Codes/pyamrex#392] [Q&A] [answered]
As titled, I didn't find any relevant codes in the manual about adaptive mesh refinement. It seems that both two example codes (HeatEquation.py and MultiFab.py) do not enable this function at all. I wonder whether there is a python code involving adaptive mesh refinement for reference. Thx!

**Comment by WeiqunZhang** [comment]
Not everything in amrex is ported to python. For an AMR code, you will have to write in C++ first. Then you can add some wrappers that allow you to drive your code from python.

**Reply by ax3l** [reply]
I agree with @WeiqunZhang for the current state.

One can certainly set up a lot for a MR simulation already from Python, e.g., setting the `amr.max_level` via `amrex.ParmParse.add` and likewise for `amr.n_cell`.

A couple of helper functions, e.g., from `amrex::AmrCore` to mark the cells for refinement are a bit hard to expose to Python right now due to their `virtual` nature, but once one has derived a C++ simulation class from them once could go back to Python and implement details of `AmrCore::ErrorEst` and `AmrCore::MakeNewLevelFromScratch` to mark and allocate cells for refinement from Python.

Update: for the last paragraph, we could actually expose `AmrCore` in pyAMReX ([like this](https://pybind11.readthedocs.io/en/stable/advanced/classes.html#overriding-virtual-functions-in-python)) and let users implement their simulation class fully in Python.

**Reply by Simon067** [reply]
has this happened? Can simulation classes now be fully implemented in python?

**Reply by ax3l** [reply, marked as answer]
Yes, AMRCore and AMRMesh are exposed now as well and you can build standalone AMReX apps just like one would in C++

https://pyamrex.readthedocs.io/en/latest/usage/api.html#amrcore

Basic needs: #582 
Small follow-up on completeness: #584 (tested when implementing the AMReX tutorials in pure Python)
